crates/bevy_transform/src/systems.rs

```diff
@@ -46,21 +46,29 @@ pub fn propagate_transforms(
             changed |= children_changed;
 
             for child in children.iter() {
-                propagate_recursive(
-                    &global_transform,
-                    &transform_query,
-                    &parent_query,
-                    &children_query,
-                    entity,
-                    *child,
-                    changed,
-                );
+                // SAFETY: Assuming the hierarchy is consistent, we can be sure that each `child` entity
```

Comment:
Thinking on this a bit more. I'm still not satisfied with this safety comment. We can't assume that the hierarchy is consistent and tree/forest-like. That's what the assertion in the recursive function is for. Not sure what the best wording here is, since the validation for the safety guarantees is asserted (recursively) in the called function itself, but is only safe under the assumption that individual entire trees are not aliased.

Reply:
We could document `propagate_recursive` with the invariant that it will panic if the hierarchy is malformed, and then cite that invariant in this safety comment. Does that seem right?

Reply:
We should definitely document the panic on `propagate_recursive`, but  the safety justification here is only valid if *both* the uniqueness of the roots AND the panic are included. The safety guarantee requires both the hierarchy to be not malformed (or panic if it is), and for the tree from the root down to be uniquely accessed from a single thread for it to be valid.

Reply:
I ended up rewriting this comment and the docs for `propagate_recursive` entirely.